src/scheduler/mod.rs

```diff
@@ -232,13 +232,35 @@ impl<'a, R: 'a + Resource + Send + Sync> Scheduler<R> {
             .lock()
             .unwrap()
             .new_ident(priority, name);
-        self.scheduler_root.lock().unwrap().schedule(
-            task_ident,
-            task,
-            resources.clone(),
-            self.poll_interval,
-            self.control_chan.clone(),
-        )
+
+        self.scheduler_root
+            .lock()
+            .unwrap()
+            .start_scheduling(task_ident.clone(), task);
+
+        for resource in resources.iter() {
+            if self
+                .scheduler_root
+                .lock()
+                .unwrap()
+                .enqueue_task_for_resource(&task_ident, resource)?
+            {
+                break;
+            };
+
+            // Give the just-enqueued task time to be assigned to a resource – so a less-preferred resource is not
+            // assigned if a more-preferred resource is available. Note that we are not holding a lock on
+            // `self.scheduler_root`, so the scheduler loop will have an opportunity to work, if possible.
+
+            thread::sleep(self.poll_interval);
+        }
+
+        self.scheduler_root
+            .lock()
+            .unwrap()
+            .finish_scheduling(task_ident);
```

Comment:
Shouldn't we do this when `Command::Finished` is sent for that particular `task_ident`? From my understanding, although we have enqueued the task to all of the resources, it doesn't necessarrily mean that the task is definitely assigned. Right?

Reply:
That said, we could also hold onto it a little longer and only remove from `assigned_tasks` when the task has finished performing (as you suggest). This might make that bookkeeping more valuable, and I don't think it harms anything. My point here is that it's main purpose for existing is only for scheduling, so it can be safely removed now.

Tangentially, this discussion raises the point that we probably need to also consider the case in which a performed task never completes cleanly and have a mechanism for ensuring cleanup anyway. (This can be added later, though.)

Reply:
Got it, thanks.

Reply:
I convinced myself the marginal value of extending the semantics of `assigned_tasks` for long-term tracking is worth it. The `finish_scheduling` mechanism was a logical step of the refactor, but since it wasn't needed for any other purpose, it can be removed now. My latest commit does this. Thanks.